### PR TITLE
(DOCSP-39838) [CPP Driver] Write Data to MongoDB > Replace a Document

### PR DESCRIPTION
**Pull request by elyse-mdb**
# Pull Request Info
Added a new page documenting the replace_one() method. 

[PR Reviewing Guidelines](https://github.com/mongodb/docs-java/blob/master/REVIEWING.md)

JIRA - [DOCSP-39838](https://jira.mongodb.org/browse/DOCSP-39838)
Staging - 
- [Replace Documents](https://deploy-preview-62--docs-cpp.netlify.app/write/replace/) 
- [Write Data to MongoDB](https://deploy-preview-62--docs-cpp.netlify.app/write/#replace-one)
## Self-Review Checklist

- [ ] Is this free of any warnings or errors in the RST?
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Are all the links working?
- [x] Are the [facets and meta keywords](https://wiki.corp.mongodb.com/display/DE/Docs+Taxonomy) accurate?

### CODE REVIEW

File: source/write/replace.txt [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+
+   The values of ``_id`` fields are immutable. If your replacement document specifies a value for the ``_id`` field, 
+   it must match the ``_id`` value of the existing document.
+
+Replace One Document Example
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The following example uses the ``replace_one()`` method to replace 
+a document that has a ``name`` field value of ``Nobu`` with a new document 
+that has a ``name`` field value of ``La Bernadin``:
```
Comment (norareidy) [inline source/write/replace.txt:70]: S: when describing fields and values, we don't use quotes for field names (so you did that correctly!) but we use quotes for values:
```suggestion
a document that has a ``name`` field value of ``"Nobu"`` with a new document 
that has a ``name`` field value of ``"La Bernadin"``:
```

File: source/write/replace.txt [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+
+.. literalinclude:: /includes/write/replace.cpp
+   :start-after: start-replace-one
+   :end-before: end-replace-one
+   :language: cpp
+   :dedent:
+
+To check if you successfully replaced the document, you can use the :ref:`find_one() <cpp-retrieve-find-one>` method to print out the new document:
```
Comment (norareidy) [inline source/write/replace.txt:78]: S: we try to avoid adding a link to a term without describing where the link goes ([style guide](https://www.mongodb.com/docs/meta/style-guide/style/links-and-cross-references/#make-link-text-clear)). You could remove this inline link and add a tip admonition after the code example that says something like "To learn more about the ``find_one()`` method, see :ref:`cpp-retrieve-find-one` in the Retrieve Data guide."

File: source/write/replace.txt [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+      
+      New document: { "_id" : { "$oid" : "5eb3d668b31de5d588f42e34" }, "name" : "La Bernadin" }  
+
+Options 
+~~~~~~~
+
+You can modify the behavior of the ``replace_one()`` method by
+passing an instance of the ``mongocxx::options::replace`` class as an optional
+argument. The following table describes the fields you can set in a
+``mongocxx::options::update`` instance:
```
Comment (norareidy) [inline source/write/replace.txt:100]: I: I think this should say mongocxx::options::replace:
```suggestion
``mongocxx::options::replace`` instance:
```

File: source/write/replace.txt [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+       - :dbcommand:`currentOp` output, in the :data:`currentOp.command
+         <currentOp.command>` ``comment`` field
+
+       | For more information, see the :manual:`insert Command
+         Fields guide </reference/command/insert/#command-fields>` in the {+mdb-server+} manual.
+
+   * - ``hint``
+     - | Specifies the index to scan for documents that match the query filter.
+         For more information, see the
+         :manual:`Update Statement fields guide </reference/command/update/#std-label-update-statement-documents>`
```
Comment (norareidy) [inline source/write/replace.txt:135]: S: you can link directly to "hint" and make the link text a little more specific:
```suggestion
         :manual:`hint field </reference/command/update/#std-label-update-command-hint>`
```

File: source/write/replace.txt [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+     - | Specifies whether the replace operation performs an upsert operation if no 
+         documents match the query filter.
+       | Defaults to ``false``.
+
+   * - ``write_concern``
+     - | Sets the write concern for the operation.
+         For more information, see :manual:`Write Concern </reference/write-concern/>`
+         in the {+mdb-server+} manual.
+
+Example: ``hint``
```
Comment (norareidy) [inline source/write/replace.txt:156]: S: the [style guide](https://www.mongodb.com/docs/meta/style-guide/style/titles-and-headings/#style-and-structure) doesn't like monospace in headers, so I'd remove that:
```suggestion
Example: hint
```

File: source/write/replace.txt [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+^^^^^^^^^^^^^^^^^
+
+The following example uses the :ref:`create_index() <cpp-indexes>` method to create an ascending single-field index 
+on the ``name`` field. It then instantiates a ``mongocxx::options::replace`` object called ``opts`` and sets its ``hints`` field value to the name 
+of the new index, ``name_1``. When ``opts`` is passed in as an optional argument, the ``replace_one()`` method searches the ``name_1`` index for a
+``name`` field value of ``Nobu`` and replaces the associated document with a new document that has a ``name`` field value of ``La Bernadin``.
```
Comment (norareidy) [inline source/write/replace.txt:162]: S: I think you can condense this description a little bit. We try to describe what's happening in the code conceptually more than describing the actual code text (names of variables, for example), so you could edit to something like:
```suggestion
The following example uses the ``create_index()`` method to create an ascending single-field index 
on the ``name`` field. It then passes a ``mongocxx::options::replace`` object to the ``replace_one()`` method and sets its ``hint`` field to the new index. This instructs the operation to use the ``name`` field index when replacing the document with a ``name`` value of ``"Nobu"``:
```

Comment (norareidy) [reply source/write/replace.txt:162]: And since you have the link below to the Indexes guide, I don't think you need the inline link on create_index()

File: source/write/replace.txt [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+
+.. literalinclude:: /includes/write/replace.cpp
+   :start-after: start-replace-options-hint
+   :end-before: end-replace-options-hint
+   :language: cpp
+   :dedent:
+
+To learn more about indexes, see :ref:`Optimize Queries with Indexes <cpp-indexes>`.
+
+Example: ``upsert``
```
Comment (norareidy) [inline source/write/replace.txt:172]: S: same suggestion about monospace
```suggestion
Example: upsert
```

File: source/write/replace.txt [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+To learn more about indexes, see :ref:`Optimize Queries with Indexes <cpp-indexes>`.
+
+Example: ``upsert`` 
+^^^^^^^^^^^^^^^^^^^
+
+The following example instantiates a ``mongocxx::options::replace`` object called ``opts``
+and sets its ``upsert`` field value to ``true``.  When ``opts`` is passed in as an optional argument, the ``replace_one()`` method
+searches the ``sample_restaurants.restaurants`` collection for a document with a ``name`` field value of ``In-N-Out Burger``. 
+Because this sample collection does not contain a document matching this query filter,
+the ``replace_one()`` method upserts a new document with a ``name`` field value of ``Shake Shack`` into the collection.
```
Comment (norareidy) [inline source/write/replace.txt:179]: S: similar to my suggestion above, I'd shorten this description to something resembling:
```suggestion
The following example passes a ``mongocxx::options::replace`` object to the ``replace_one()`` method and sets its ``upsert`` field value to ``true``.  Because no documents match the query filter, this instructs the operation to insert a new document with a ``name`` field value of ``"Shake Shack"`` into the collection:
```

File: source/write/replace.txt [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+
+The ``replace_one()`` method returns an instance of
+the ``mongocxx::result::replace`` class. This class contains the following
+member functions:
+
+.. list-table::
+   :widths: 30 70
+   :header-rows: 1
+
+   * - Field
```
Comment (norareidy) [inline source/write/replace.txt:207]: S: I'd change to "function" or "method":
```suggestion
   * - Function
```

File: source/write/replace.txt [hunk 10]
```diff
@@ -0,0 +1,282 @@
...
+         count.
+         
+   * - ``result()``
+     - | Returns the bulk write result for the operation.
+
+   * - ``upserted_id()``
+     - | Returns the ID of the document that was upserted in the database, if the driver
+         performed an upsert.
+
+Example: ``matched_count()``
```
Comment (norareidy) [inline source/write/replace.txt:226]: S:
```suggestion
Example: matched_count()
```

File: source/write/replace.txt [hunk 11]
```diff
@@ -0,0 +1,282 @@
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The following example uses the ``replace_one()`` method to replace a document that has
+a ``name`` field value of ``Shake Shack`` with a new document that has a ``name`` 
+field value of ``In-N-Out Burger``. It then calls the ``matched_count()`` member 
+function to print the number of documents that match the search for a ``name`` field value of ``Shake Shack``:
```
Comment (norareidy) [inline source/write/replace.txt:232]: S: I think you can just say "query filter" here
```suggestion
function to print the number of documents that match the query filter:
```

File: source/write/replace.txt [hunk 12]
```diff
@@ -0,0 +1,282 @@
...
+      :language: cpp
+      :dedent:
+
+   .. output:: 
+      :language: cli
+      :visible: false
+
+      Matched documents: 11
+
+Example: ``upserted_id()``
```
Comment (norareidy) [inline source/write/replace.txt:248]: S:
```suggestion
Example: upserted_id()
```

File: source/write/replace.txt [hunk 13]
```diff
@@ -0,0 +1,282 @@
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The following example uses the ``replace_one()`` method to find and replace a document that has a ``name`` field value of ``In-N-Out Burger``. 
+Because the upsert option is set to true, the {+driver-short+} inserts a new document when the query filter doesn't match any existing documents. 
+It then cals the ``upserted_id()`` member function to print the ``_id`` field value of the upserted document.
```
Comment (norareidy) [inline source/write/replace.txt:253]: S: Some small formatting/wording suggestions:

```suggestion
The following example uses the ``replace_one()`` method to replace a document that has a ``name`` field value of ``"In-N-Out Burger"``. 
Because the ``upsert`` option is set to ``true``, the {+driver-short+} inserts a new document when the query filter doesn't match any existing documents. Then, the code
calls the ``upserted_id()`` member function to print the ``_id`` field value of the upserted document:
```

File: source/write/replace.txt [hunk 14]
```diff
@@ -0,0 +1,282 @@
...
+   .. input:: /includes/write/replace.cpp
+      :start-after: start-replace-result-upsert
+      :end-before: end-replace-result-upsert
+      :language: cpp
+      :dedent:
+
+   .. output:: 
+      :language: cli
+      :visible: false
+
```
Comment (norareidy) [inline source/write/replace.txt:266]: S: since the user's ID value will be different, I'd add a comment here that says "your value may differ" or something like that
```suggestion
     // Your ID value may differ
```

Comment (norareidy) [reply source/write/replace.txt:266]: The code output [here](https://www.mongodb.com/docs/drivers/rust/current/usage-examples/updateMany/#example) in the Rust docs does something similar

**Review by norareidy — CHANGES_REQUESTED** [review]
Great work on this page! I left some suggestions/comments, mostly small ones

File: source/write/replace.txt [hunk 15]
```diff
@@ -0,0 +1,283 @@
...
+      :language: cpp
+      :dedent:
+   
+   .. output:: 
+      :language: cli
+      :visible: false
+      
+      New document: { "_id" : { "$oid" : "5eb3d668b31de5d588f42e34" }, "name" : "La Bernadin" }  
+
+To learn more about the find_one() method, see :ref:`cpp-retrieve-find-one` in the Retrieve Data guide.
```
Comment (norareidy) [inline source/write/replace.txt:94]: S:
```suggestion
To learn more about the ``find_one()`` method, see :ref:`cpp-retrieve-find-one` in the Retrieve Data guide.
```

File: source/write/replace.txt [hunk 16]
```diff
@@ -0,0 +1,283 @@
...
+     - | Specifies the index to scan for documents that match the query filter.
+         For more information, see the :manual:`hint field </reference/command/update/#std-label-update-command-hint>`
+         in the {+mdb-server+} manual. 
+
+   * - ``let``
+     - | Specifies a document containing variables and their values to be used in the ``replace_one()`` method. 
+         This allows you to improve code readability by separating the variables from the operation text. 
+         Values must be constant or closed expressions that don't reference document fields. 
+         For more information, see the 
+         :manual:`update Command Fields guide </reference/command/update/#std-label-replace-let-syntax>`
```
Comment (norareidy) [inline source/write/replace.txt:144]: S: I think this link is slightly off, so it's just linking to the main page; you should be able to use this link which goes directly to the let field:
```suggestion
         :manual:`let field </reference/command/update/#std-label-update-let-syntax>`
```

File: source/write/replace.txt [hunk 17]
```diff
@@ -0,0 +1,283 @@
...
+method after setting its ``hint`` field to the new index. This instructs the replace operation to search the ``name`` 
+field index when replacing a document that has a ``name`` field value of ``"Nobu"``:
+
+.. literalinclude:: /includes/write/replace.cpp
+   :start-after: start-replace-options-hint
+   :end-before: end-replace-options-hint
+   :language: cpp
+   :dedent:
+
+To learn more about indexes, see :ref:`Optimize Queries with Indexes <cpp-indexes>`.
```
Comment (norareidy) [inline source/write/replace.txt:171]: S: [nit] you can be a little more specific about where this link is going
```suggestion
To learn more about indexes, see the :ref:`Optimize Queries with Indexes <cpp-indexes>` guide.
```

File: source/write/replace.txt [hunk 18]
```diff
@@ -0,0 +1,283 @@
...
+
+Example: upsert 
+^^^^^^^^^^^^^^^^^^^
```
Comment (norareidy) [inline source/write/replace.txt:174]: S: I'd clarify that this is an option in the header. Also, make sure your header underline is the same length as the header text
```suggestion
Example: upsert Option
^^^^^^^^^^^^^^^^^^^^^^
```

Comment (norareidy) [reply source/write/replace.txt:174]: Applies to the hint example as well

File: source/write/replace.txt [hunk 19]
```diff
@@ -0,0 +1,283 @@
...
+         
+   * - ``result()``
+     - | Returns the bulk write result for the operation.
+
+   * - ``upserted_id()``
+     - | Returns the ID of the document that was upserted in the database, if the driver
+         performed an upsert.
+
+Example: matched_count()
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (norareidy) [inline source/write/replace.txt:227]: S: make sure the underline is the same length as the text; applies to all

File: source/write/replace.txt [hunk 20]
```diff
@@ -0,0 +1,283 @@
...
+      :language: cli
+      :visible: false
+
+      // Your ID value may differ
+      Upserted ID: 67128c5ecc1f8c15ea26fcf8
+
+Additional Information 
+----------------------
+
+To learn more about creating query filters, see :ref:`cpp-specify-query`.
```
Comment (norareidy) [inline source/write/replace.txt:273]: S: clarify where this link is going
```suggestion
To learn more about creating query filters, see the :ref:`cpp-specify-query` guide.
```

File: source/includes/write/replace.cpp [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+    mongocxx::client client(uri);
+
+    // start-db-coll
+    auto db = client["sample_restaurants"];
+    auto collection = db["restaurants"];
+    // end-db-coll
+
+    {
+        // replaces the fields and values of a document with the name "Bagels N Buns" with a document 
+        // with the same _id value and no fields other than the name "2 Bagels 2 Buns"
```
Comment (norareidy) [inline source/includes/write/replace.cpp:24]: S: you can shorten this comment; also, make sure it mentions the correct values
```suggestion
        // Replaces a document that has a "name" value of "Nobu" with a document that has a "name" of "La Bernadin"
```

File: source/includes/write/replace.cpp [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+        opts.hint(mongocxx::hint{"name_1"});
+
+        auto query_filter = make_document(kvp("name", "Nobu"));
+        auto replace_doc = make_document(kvp("name", "La Bernadin"));
+
+        auto result = collection.replace_one(query_filter.view(), replace_doc.view(), opts);
+        // end-replace-options-hint 
+    }
+    
+    {
```
Comment (norareidy) [inline source/includes/write/replace.cpp:56]: S: add a comment, something like:
```suggestion
    // Replaces a document that has a "name" value of "In-N-Out Burger" and instructs the operation to insert a new document if none match
    {
```

**Review by norareidy — CHANGES_REQUESTED** [review]
Almost there, just a few more things!

File: source/write/replace.txt [hunk 23]
```diff
@@ -0,0 +1,283 @@
...
+:atlas:`Get Started with Atlas </getting-started>` guide.
+
+Replace Operation
+-----------------
+
+You can perform a replace operation by calling the ``replace_one()`` method.
+This method removes all fields except the ``_id`` field from the first document that
+matches the search criteria. It then inserts the fields and values you specify into the document. 
+
+Each replace method requires the following parameters:
```
Comment (norareidy) [inline source/write/replace.txt:51]: S: since there's only one replace method, I'd reword slightly
```suggestion
The ``replace_one()`` method requires the following parameters:
```

File: source/includes/write/replace.cpp [hunk 24]
```diff
@@ -0,0 +1,101 @@
...
+        // Replaces a document that has a "name" value of "Nobu" and instructs the operation to use the "name_1" field index
```
Comment (norareidy) [inline source/includes/write/replace.cpp:57]: S: I'd change this either to the "name" field index or the "name_1" index (since it's an index on the name field, but the index itself is called name_1)
```suggestion
        // Replaces a document that has a "name" value of "Nobu" and instructs the operation to use the "name" field index
```

**Review by norareidy — APPROVED** [review]
LGTM with a couple small suggestions - great job!!

File: source/includes/write/replace.cpp [hunk 25]
```diff
@@ -0,0 +1,101 @@
...
+
+        auto query_filter = make_document(kvp("name", "Nobu"));
+        auto replace_doc = make_document(kvp("name", "La Bernadin"));
+
+        auto result = collection.replace_one(query_filter.view(), replace_doc.view(), opts);
+        // end-replace-options-hint 
+    }
+    
+    {
+        // Replaces a document that has a "name" value of "Nobu" and instructs the operation to use the "name" field index
```
Comment (kevinAlbs) [inline source/includes/write/replace.cpp:57]:
```suggestion
        // Replaces a document that has a "name" value of "In-N-Out Burger" and instructs the operation
        // to insert a new document if none match.
```

**Review by kevinAlbs — APPROVED** [review]
LGTM with a minor non-public-facing comment fix.
